File: apps/desktop/src-tauri/src/agent/protocol.rs
//! §26 presentation envelope, mirrored for the Rust host (contract §48:
//! JSONL protocol deserialize). The Rust host forwards envelopes verbatim —
//! it never interprets business payloads, only classifies
//! requestId-present = response / requestId-absent = streaming event.

use serde::{Deserialize, Serialize};
use serde_json::{json, Value};

/// Canonical envelope of UI Spec §26.1.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct Envelope {
    #[serde(default = "default_protocol_version")]
    pub protocol_version: u32,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub request_id: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub session_id: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub turn_id: Option<i64>,
    pub seq: i64,
    #[serde(rename = "type")]
    pub type_: String,
    pub data: Value,
}

fn default_protocol_version() -> u32 {
    1
}

impl Envelope {
    pub fn is_response(&self) -> bool {
        self.request_id.is_some()
    }

    #[allow(dead_code)] // classification helper; consumed by host tests
    pub fn event_type(&self) -> &str {
        &self.type_
    }

    pub fn parse(line: &str) -> Result<Envelope, serde_json::Error> {
        serde_json::from_str(line)
    }

    /// Uniform failure data shape `{ ok:false, error:{ code, message } }`.
    pub fn error_data(code: &str, message: &str) -> Value {
        json!({ "ok": false, "error": { "code": code, "message": message } })
    }

    /// A host-originated envelope (no requestId → published as an event).
    pub fn host_event(type_: &str, seq: i64, data: Value) -> Envelope {
        Envelope {
            protocol_version: 1,
            request_id: None,
            session_id: None,
            turn_id: None,
            seq,
            type_: type_.to_string(),
            data,
        }
    }

    /// A synthetic failure response (used when the sidecar died mid-request).
    pub fn failure_response(request_type: &str, seq: i64, code: &str, message: &str) -> Envelope {
        Envelope {
            protocol_version: 1,
            request_id: None,
            session_id: None,
            turn_id: None,
            seq,
            type_: request_type.to_string(),
            data: Envelope::error_data(code, message),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_a_bridge_response_line_verbatim() {
        let line = r#"{"protocolVersion":1,"requestId":"req-1","type":"session/created","sessionId":"session-abc","seq":12,"data":{"session":{"sessionId":"session-abc","title":"","createdAt":1700000000000,"messageCount":0}}}"#;
        let env = Envelope::parse(line).expect("valid envelope");
        assert_eq!(env.protocol_version, 1);
        assert_eq!(env.request_id.as_deref(), Some("req-1"));
        assert_eq!(env.type_, "session/created");
        assert_eq!(env.session_id.as_deref(), Some("session-abc"));
        assert_eq!(env.seq, 12);
        assert!(env.is_response());
        assert_eq!(env.data["session"]["messageCount"], 0);
    }

    #[test]
    fn parses_null_session_and_turn_fields() {
        // The bridge writes explicit nulls on pure events.
        let line = r#"{"protocolVersion":1,"type":"runtime/status","sessionId":null,"turnId":null,"seq":3,"data":{"status":"ready"}}"#;
        let env = Envelope::parse(line).expect("valid envelope");
        assert!(!env.is_response());
        assert_eq!(env.session_id, None);
        assert_eq!(env.turn_id, None);
        assert_eq!(env.type_, "runtime/status");
        assert_eq!(env.data["status"], "ready");
    }

    #[test]
    fn classifies_streaming_turn_events_as_events() {
        let turn_start = Envelope::parse(
            r#"{"protocolVersion":1,"type":"turn/start","sessionId":"s1","turnId":1,"seq":41,"data":{"userText":"hi"}}"#,
        )
        .unwrap();
        let chunk = Envelope::parse(
            r#"{"protocolVersion":1,"type":"assistant/chunk","sessionId":"s1","turnId":1,"seq":55,"data":{"turnId":1,"text":"你"}}"#,
        )
        .unwrap();
        assert!(!turn_start.is_response());
        assert!(!chunk.is_response());
        assert_eq!(turn_start.turn_id, Some(1));
        // turn/end reason stays opaque data — the host never decodes it.
        assert_eq!(turn_start.data["userText"], "hi");
    }

    #[test]
    fn serde_round_trip_keeps_field_names_camel_case() {
        let env = Envelope::host_event("runtime/status", 1, json!({ "status": "disconnected" }));
        let text = serde_json::to_string(&env).unwrap();
        assert!(text.contains("\"protocolVersion\":1"), "got {text}");
        assert!(text.contains("\"runtime/status\""), "got {text}");
        let back: Envelope = serde_json::from_str(&text).unwrap();
        assert_eq!(back, env);
    }

    #[test]
    fn failure_response_mirrors_the_uniform_error_shape() {
        let env = Envelope::failure_response("turn.send", 0, "NOT_READY", "Agent 未运行");
        assert_eq!(env.event_type(), "turn.send");
        assert_eq!(env.data["ok"], false);
        assert_eq!(env.data["error"]["code"], "NOT_READY");
    }
}

File: apps/agent-bridge/src/protocol.ts
/**
 * Desktop presentation protocol — the exact wire contract of UI Spec §26.
 *
 * Generic agent presentation protocol shared (by text) between the Node
 * Agent Bridge and the React frontend. NO Chinook-specific event types and
 * NO `@deepseek-ai/*` types may ever leak in here: arguments/result JSON is
 * opaque to the wire (UI Spec §26.5).
 *
 * Every protocol message is one JSON line on the bridge's stdout:
 *
 * ```ts
 * interface Envelope {
 *   protocolVersion: 1;
 *   requestId?: string;      // present on requests and their responses
 *   sessionId?: string | null;
 *   turnId?: number | null;  // dsh turn number within the session
 *   seq: number;             // per-session dsh event seq, else presentation counter
 *   type: string;            // see below
 *   data: unknown;           // JSON-serializable payload
 * }
 * ```
 *
 * This module contains types + tiny helpers only; importing it must have no
 * runtime side effects beyond these constants.
 */

export const PROTOCOL_VERSION = 1 as const;

// ---------------------------------------------------------------------------
// §26.1 envelope
// ---------------------------------------------------------------------------

export interface Envelope {
  protocolVersion: typeof PROTOCOL_VERSION;
  requestId?: string;
  sessionId?: string | null;
  turnId?: number | null;
  seq: number;
  type: string;
  data: unknown;
}

/** Uniform error response data (mirrors the V1 tool contract deliberately). */
export interface ProtocolErrorData {
  ok: false;
  error: { code: string; message: string };
}

export type RuntimeStatus = "starting" | "ready" | "restoring" | "disconnected" | "error" | "restarting";
export type TurnEndReason = "completed" | "aborted" | "blocked" | "error" | "max-tokens" | "interrupted";
export type AssistantStatus = "complete" | "stopped" | "failed" | "limited" | "interrupted" | "no-answer";

// ---------------------------------------------------------------------------
// §26.2 request payloads (received on stdin / from Tauri Commands)
// ---------------------------------------------------------------------------

export interface SessionListRequest {
  type: "session.list";
}
export interface SessionCreateRequest {
  type: "session.create";
}
export interface SessionOpenRequest {
  type: "session.open";
  sessionId: string;
}
export interface AgentStatusRequest {
  type: "agent.status";
}
export interface AgentRestartRequest {
  type: "agent.restart";
}
export interface TurnSendRequest {
  type: "turn.send";
  sessionId: string;
  text: string;
}

/**
 * Model-endpoint configuration. `data` is deliberately absent from get/test —
 * neither takes an argument, so neither can carry a credential value back or
 * forth. `config.save` carries a key ONE WAY only: the wire never returns one.
 */
export interface ConfigGetRequest {
  type: "config.get";
}
export interface ConfigSaveRequest {
  type: "config.save";
  data: {
    /** "" clears the override and restores the endpoint default. */
    baseUrl?: string;
    model?: string;
    /** Absent or "" keeps the stored key. */
    apiKey?: string;
    /** Explicit removal of the stored key. */
    clearApiKey?: boolean;
  };
}
export interface ConfigTestRequest {
  type: "config.test";
}

export type Request =
  | SessionListRequest
  | SessionCreateRequest
  | SessionOpenRequest
  | AgentStatusRequest
  | AgentRestartRequest
  | TurnSendRequest
  | ConfigGetRequest
  | ConfigSaveRequest
  | ConfigTestRequest;

// ---------------------------------------------------------------------------
// §26.4 session & item shapes (React-consumable; no DSH types)
// ---------------------------------------------------------------------------

export interface SessionSummary {
  sessionId: string;
  title: string;
  createdAt: number;
  messageCount?: number;
}

export interface ToolActivity {
  name: string; // wire name (search_catalog, …)
  callId: string;
  arguments: Record<string, unknown>;
  ok: boolean;
  error?: { code: string; message: string };
  durationMs?: number;
  result?: unknown; // parsed structured result (or raw JSON string)
  ts: number;
}

export type ConversationItem =
  | {
      kind: "user";
      id: string;
      text: string;
      ts: number;
    }
  | {
      kind: "assistant";
      id: string; // `${sessionId}:${turnId}`
      text: string;
      ts: number;
      status: AssistantStatus;
      tools: ToolActivity[]; // ordered per turn; empty when none
      // Reducer-side superset (§28.2): recorded on the open item during a live
      // turn / carried by snapshot items so resumed failures keep their caption.
      error?: { code: string; message: string };
      usage?: unknown;
      interrupted?: boolean;
      archived?: { olderCount: number };
    };

/** §28.1 — one row per presentation event; drawer groups 'chunk' rows visually. */
export type TimelineRow =
  | { type: "turn/start"; turnId: number; ts: number }
  | { type: "tool/call" | "tool/result"; turnId: number; ts: number; tool: ToolActivity }
  | { type: "chunk"; turnId: number; ts: number; chars: number } // per assistant/chunk
  | { type: "step"; turnId: number; ts: number; usage?: unknown; interrupted?: boolean } // assistant/message
  | { type: "turn/error"; turnId: number; ts: number; code: string; message: string }
  | { type: "turn/end"; turnId: number; ts: number; reason: TurnEndReason };

// ---------------------------------------------------------------------------
// §26.2/§26.3 response data shapes
// ---------------------------------------------------------------------------

export interface RuntimeStatusData {
  status: RuntimeStatus;
  detail?: string;
}
export interface SessionListData {
  sessions: SessionSummary[];
}
export interface SessionCreatedData {
  session: SessionSummary; // title may update later
}
export interface SessionOpenedData {
  session: SessionSummary;
  items: ConversationItem[];
  log: TimelineRow[];
}
export interface SessionTitleUpdatedData {
  sessionId: string;
  title: string;
}
export interface TurnStartData {
  userText: string;
}
export interface ToolCallData {
  turnId: number;
  tool: { name: string; arguments: Record<string, unknown>; callId: string };
}
export interface ToolResultData {
  turnId: number;
  tool: {
    name: string;
    callId: string;
    durationMs: number;
    ok: boolean;
    error?: { code: string; message: string };
    result?: unknown;
  };
}
export interface AssistantChunkData {
  turnId: number;
  text: string; // text only; the bridge dropped reasoning
}
export interface AssistantMessageData {
  turnId: number;
  text: string;
  usage?: unknown;
  interrupted?: boolean;
}
export interface TurnErrorData {
  turnId: number;
  code: string;
  message: string;
}
export interface TurnEndData {
  turnId: number;
  reason: TurnEndReason;
}
export interface AgentStatusData {
  status: RuntimeStatus;
}
export interface TurnSendAcceptData {
  accepted: true;
}

/**
 * §26.2 model-endpoint configuration rows.
 *
 * The security-relevant shape here is what is NOT in it: `ApiConfigData` has
 * no field that can carry a credential value. The UI sends a key and reads
 * back only whether one resolves, from which layer, and whether this process
 * may change it (contract §44 — React never receives the model credential).
 */
export interface ApiKeyStateData {
  /** The credential reference in use. */
  ref: string;
  configured: boolean;
  /** Supplying layer; `env` means inherited and therefore read-only. */
  source?: string;
  writable: boolean;
}
export interface ApiConfigData {
  provider: string;
  model: string;
  /** User-supplied base URL; "" means the default endpoint is in use. */
  baseUrl: string;
  /** Whether the user layer owns `baseURL`. */
  baseUrlOverridden: boolean;
  apiKey: ApiKeyStateData;
}
export interface ApiConfigSavedData {
  config: ApiConfigData;
  /** True when the default model changed and the live session was rebound. */
  modelChanged: boolean;
  /** A credential write the store refused; the settings write still applied. */
  apiKeyError?: string;
}
/**
 * The probe's OWN outcome. The success flag is `connected`, never `ok`:
 * `{ ok: false }` is the protocol's error-envelope discriminator
 * (`isErrorData`), so a successful response carrying a negative verdict under
 * that name would be read as a failed request by every layer downstream.
 */
export interface ApiConfigTestedData {
  connected: boolean;
  /** Machine-routable failure code; absent on success. */
  code?: string;
  /** Render-ready Chinese caption; "" on success. */
  message: string;
  latencyMs: number;
}

/** type → response/event data payload map (mirrors §26.2–26.3 rows). */
export interface DataByType {
  "runtime/status": RuntimeStatusData;
  "session/list": SessionListData;
  "session/created": SessionCreatedData;
  "session/opened": SessionOpenedData;
  "session/title": SessionTitleUpdatedData;
  "turn/start": TurnStartData;
  "tool/call": ToolCallData;
  "tool/result": ToolResultData;
  "assistant/chunk": AssistantChunkData;
  "assistant/message": AssistantMessageData;
  "turn/error": TurnErrorData;
  "turn/end": TurnEndData;
  "agent.status": AgentStatusData;
  "turn.send": TurnSendAcceptData;
  "session.list": SessionListData;
  "session.create": SessionCreatedData;
  "session.open": SessionOpenedData;
  "config.describe": ApiConfigData;
  "config.saved": ApiConfigSavedData;
  "config.tested": ApiConfigTestedData;
}

/** Request names whose canonical success envelope type differs (response-as-event). */
export type RequestType = Request["type"];

/** Request → success response envelope type (reducer rows consume the canonical type). */
export const SUCCESS_TYPE_BY_REQUEST: Record<RequestType, string> = {
  "session.list": "session/list",
  "session.create": "session/created",
  "session.open": "session/opened",
  "agent.status": "agent.status",
  "agent.restart": "agent.restart",
  "turn.send": "turn.send",
  "config.get": "config.describe",
  "config.save": "config.saved",
  "config.test": "config.tested",
};

/** Error code vocabulary (spec §26.2 + bridge-defined generic codes). */
export const ERROR_CODES = {
  TURN_ACTIVE: "TURN_ACTIVE",
  NOT_READY: "NOT_READY",
  NOT_FOUND: "NOT_FOUND",
  INVALID_ARGUMENT: "INVALID_ARGUMENT",
  UNKNOWN_REQUEST: "UNKNOWN_REQUEST",
  NOT_IMPLEMENTED: "NOT_IMPLEMENTED",
  BOOT_FAILED: "BOOT_FAILED",
  /** The profile did not mount a service the configuration surface needs. */
  CONFIG_UNAVAILABLE: "CONFIG_UNAVAILABLE",
} as const;

export function errorData(code: string, message: string): ProtocolErrorData {
  return { ok: false, error: { code, message } };
}

// ---------------------------------------------------------------------------
// tiny pure helpers (used by the bridge, mirrored where needed by tests)
// ---------------------------------------------------------------------------

let presentationSeq = 0;
export function nextPresentationSeq(): number {
  presentationSeq += 1;
  return presentationSeq;
}

export function makeEnvelope(partial: {
  type: string;
  data: unknown;
  sessionId?: string | null;
  turnId?: number | null;
  requestId?: string;
  seq?: number;
}): Envelope {
  return {
    protocolVersion: PROTOCOL_VERSION,
    requestId: partial.requestId,
    sessionId: partial.sessionId ?? null,
    turnId: partial.turnId ?? null,
    seq: partial.seq ?? nextPresentationSeq(),
    type: partial.type,
    data: partial.data,
  };
}
